modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/RegionNodeComparator.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.spi.discovery.tcp.internal;
+
+import java.util.Comparator;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * New behavior for node ordering. Sort firstly by region ID, and secondly by topology version.
+ * Node without region ID (it is equal null) is less the node with the region ID.
+ */
+public class RegionNodeComparator implements Comparator<TcpDiscoveryNode> {
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int compare(@NotNull TcpDiscoveryNode firstNode, @NotNull TcpDiscoveryNode secondNode) {
+        Long firstRegionId = firstNode.getClusterRegionId();
+        Long secondRegionId = secondNode.getClusterRegionId();
+
+        if (firstRegionId == null && secondRegionId == null)
```

Comment:
There are plenty of formatting issues - please refer to https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines